File: frontend/src/types/policyFinding.ts
export type PolicySeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'
export type PolicyCategory = 'MAINTAINABILITY' | 'CODE_QUALITY'

export interface PolicyFinding {
  id: number
  analysisRunId: number
  repositoryFullName: string
  pullRequestNumber: number
  commitSha: string
  ruleId: string
  category: PolicyCategory
  severity: PolicySeverity
  filePath: string
  lineNumber: number
  message: string
  recommendation: string
  createdAt: string
}

export interface PolicyFindingFilters {
  analysisRunId?: number
  repositoryId?: number
  severity?: PolicySeverity
  category?: PolicyCategory
  ruleId?: string
  page?: number
  size?: number
  sort?: string
}

File: backend/src/main/java/com/gatekeeper/policy/PolicyFinding.java
package com.gatekeeper.policy;

/**
 * One deterministic observation produced by a single PolicyRule. Deliberately
 * a standalone type rather than a subtype of a shared "Finding" abstraction:
 * Architecture.md's Finding hierarchy (PolicyFinding/SecurityFinding/AIFinding
 * sharing a common parent) only earns its keep once a second engine exists to
 * generalize from - introducing it now, with a single subtype, would be
 * exactly the premature abstraction SOLID warns against. Revisit this the
 * moment the Security Engine milestone begins (see the Milestone 3 writeup's
 * Future Extensibility Analysis).
 *
 * @param ruleId         the producing rule's PolicyRule#id(), for traceability
 * @param category       what kind of concern this is
 * @param severity       how urgently it needs attention
 * @param filePath       path of the file the finding was found in
 * @param lineNumber     1-based line number within that file
 * @param message        human-readable description of what was found
 * @param recommendation human-readable suggestion for resolving it
 */
public record PolicyFinding(
        String ruleId,
        PolicyCategory category,
        PolicySeverity severity,
        String filePath,
        int lineNumber,
        String message,
        String recommendation) {
}
